File: gamedex-persistence/src/main/java/com/github/ykrasik/gamedex/persistence/PersistenceServiceImpl.java
package com.github.ykrasik.gamedex.persistence;

import com.github.ykrasik.gamedex.common.service.AbstractService;
import com.github.ykrasik.gamedex.datamodel.GamePlatform;
import com.github.ykrasik.gamedex.datamodel.persistence.*;
import com.github.ykrasik.gamedex.datamodel.provider.UnifiedGameInfo;
import com.github.ykrasik.gamedex.persistence.config.PersistenceProperties;
import com.github.ykrasik.gamedex.persistence.dao.*;
import com.github.ykrasik.gamedex.persistence.entity.*;
import com.github.ykrasik.gamedex.persistence.exception.DataException;
import com.github.ykrasik.gamedex.persistence.extractor.GenreExtractor;
import com.github.ykrasik.gamedex.persistence.extractor.LibraryExtractor;
import com.github.ykrasik.gamedex.persistence.translator.exclude.ExcludedPathEntityTranslator;
import com.github.ykrasik.gamedex.persistence.translator.game.GameEntityTranslator;
import com.github.ykrasik.gamedex.persistence.translator.genre.GenreEntityTranslator;
import com.github.ykrasik.gamedex.persistence.translator.library.LibraryEntityTranslator;
import com.gs.collections.api.list.ImmutableList;
import com.gs.collections.api.list.MutableList;
import com.gs.collections.api.map.ImmutableMap;
import com.gs.collections.api.multimap.list.ImmutableListMultimap;
import com.gs.collections.impl.factory.Lists;
import com.j256.ormlite.dao.DaoManager;
import com.j256.ormlite.jdbc.JdbcPooledConnectionSource;
import com.j256.ormlite.table.TableUtils;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

import java.nio.file.Path;
import java.sql.SQLException;

/**
 * @author Yevgeny Krasik
 */
@RequiredArgsConstructor
public class PersistenceServiceImpl extends AbstractService implements PersistenceService {
    @NonNull private final PersistenceProperties properties;
    @NonNull private final GameEntityTranslator gameTranslator;
    @NonNull private final GenreEntityTranslator genreTranslator;
    @NonNull private final LibraryEntityTranslator libraryTranslator;
    @NonNull private final ExcludedPathEntityTranslator excludedPathTranslator;

    private JdbcPooledConnectionSource connectionSource;
    private GameDao gameDao;
    private GenreDao genreDao;
    private GenreGameLinkDao genreGameLinkDao;
    private LibraryDao libraryDao;
    private LibraryGameLinkDao libraryGameLinkDao;
    private ExcludePathDao excludedPathDao;

    @Override
    protected void doStart() throws Exception {
        final String dbUrl = properties.getDbUrl();
        LOG.debug("Connection url: {}", dbUrl);
        connectionSource = new JdbcPooledConnectionSource(dbUrl);
        connectionSource.setCheckConnectionsEveryMillis(properties.getConnectionCheckInterval().toMillis());

        LOG.debug("Preparing schema...");
        prepareSchema();

        LOG.debug("Instantiating DAOs...");
        instantiateDaos();
    }

    private void prepareSchema() throws SQLException {
        TableUtils.createTableIfNotExists(connectionSource, GameEntity.class);
        TableUtils.createTableIfNotExists(connectionSource, GenreEntity.class);
        TableUtils.createTableIfNotExists(connectionSource, GenreGameLinkEntity.class);
        TableUtils.createTableIfNotExists(connectionSource, LibraryEntity.class);
        TableUtils.createTableIfNotExists(connectionSource, LibraryGameLinkEntity.class);
        TableUtils.createTableIfNotExists(connectionSource, ExcludedPathEntity.class);
    }

    @SuppressWarnings("unchecked")
    private void instantiateDaos() throws SQLException {
        gameDao = DaoManager.createDao(connectionSource, GameEntity.class);
        genreDao = DaoManager.createDao(connectionSource, GenreEntity.class);
        genreGameLinkDao = DaoManager.createDao(connectionSource, GenreGameLinkEntity.class);
        ((GenreGameLinkDaoImpl)genreGameLinkDao).setDaos(gameDao, genreDao);
        libraryDao = DaoManager.createDao(connectionSource, LibraryEntity.class);
        libraryGameLinkDao = DaoManager.createDao(connectionSource, LibraryGameLinkEntity.class);
        ((LibraryGameLinkDaoImpl)libraryGameLinkDao).setDaos(gameDao, libraryDao);
        excludedPathDao = DaoManager.createDao(connectionSource, ExcludedPathEntity.class);
    }

    @Override
    protected void doStop() throws Exception {
        connectionSource.close();
    }

    @Override
    @SneakyThrows
    public Game addGame(UnifiedGameInfo gameInfo, Path path, GamePlatform platform) {
        // Insert game.
        final GameEntity game = gameTranslator.translate(gameInfo);
        game.setPath(path.toString());
        game.setPlatform(platform);
        LOG.debug("Inserting {}...", game);
        if (gameDao.create(game) != 1) {
            throw new DataException("Error inserting game: %s", gameInfo);
        }

        // Insert all new genres.
        final ImmutableList<GenreEntity> genresEntities = gameInfo.getGenres().collect(this::getOrCreateGenreByName);

        // Link genres to game.
        for (GenreEntity genre : genresEntities) {
            final GenreGameLinkEntity genreGameLinkEntity = new GenreGameLinkEntity();
            genreGameLinkEntity.setGame(game);
            genreGameLinkEntity.setGenre(genre);
            genreGameLinkDao.create(genreGameLinkEntity);
        }

        return getGameById(new Id<>(game.getId()));
    }

    @SneakyThrows
    private GenreEntity getOrCreateGenreByName(String name) {
        return genreDao.getOrCreateByName(name);
    }

    @Override
    @SneakyThrows
    public void deleteGame(Id<Game> id) {
        final int gameId = id.getId();
        LOG.debug("Deleting game with id {}...", gameId);
        if (gameDao.deleteById(gameId) != 1) {
            throw new DataException("Error deleting game: %d. Entry doesn't exist?", gameId);
        }

        final ImmutableList<GenreEntity> genres = genreGameLinkDao.getGenresByGameId(gameId);
        genreGameLinkDao.deleteByGameId(gameId);

        // Delete any genres which were only linked to this game.
        final MutableList<GenreEntity> genresToDelete = Lists.mutable.of();
        for (GenreEntity genre : genres) {
            final ImmutableList<GenreGameLinkEntity> genreLinks = genreGameLinkDao.getByGenreId(genre.getId());
            if (genreLinks.isEmpty()) {
                LOG.debug("Genre={} was only linked to gameId={}, deleting...", genre, gameId);
                genresToDelete.add(genre);
            }
        }
        if (!genresToDelete.isEmpty()) {
            genreDao.deleteIds(genresToDelete.collect(GenreEntity::getId));
        }

        libraryGameLinkDao.deleteByGameId(gameId);
    }

    @Override
    @SneakyThrows
    public ImmutableList<Game> getAllGames() {
        LOG.debug("Fetching all games...");
        final ImmutableList<GameEntity> games = Lists.immutable.ofAll(gameDao.queryForAll());

        // This is required due to lack of support for intelligent joins in OrmLite... :/
        final GenreExtractor genreExtractor = createGenreExtractor();
        final LibraryExtractor libraryExtractor = createLibraryExtractor();
        return games.collect(game -> translateGame(game, genreExtractor, libraryExtractor));
    }

    private GenreExtractor createGenreExtractor() throws SQLException {
        final ImmutableList<Genre> genres = getAllGenres();
        final ImmutableMap<Id<Genre>, Genre> genreMap = genres.groupByUniqueKey(Genre::getId);

        final ImmutableList<GenreGameLinkEntity> genreGames = genreGameLinkDao.getAll();
        final ImmutableListMultimap<Integer, GenreGameLinkEntity> genreGameMap = genreGames.groupBy(entity -> entity.getGame().getId());

        return new GenreExtractor(genreMap, genreGameMap);
    }

    private LibraryExtractor createLibraryExtractor() throws SQLException {
        final ImmutableList<Library> libraries = getAllLibraries();
        final ImmutableMap<Id<Library>, Library> libraryMap = libraries.groupByUniqueKey(Library::getId);

        final ImmutableList<LibraryGameLinkEntity> libraryGames = libraryGameLinkDao.getAll();
        final ImmutableListMultimap<Integer, LibraryGameLinkEntity> libraryGameMap = libraryGames.groupBy(entity -> entity.getGame().getId());

        return new LibraryExtractor(libraryMap, libraryGameMap);
    }

    private Game translateGame(GameEntity entity, GenreExtractor genreExtractor, LibraryExtractor libraryExtractor) {
        final ImmutableList<Genre> genres = genreExtractor.getData(entity.getId()).toList().toImmutable();
        final ImmutableList<Library> libraries = libraryExtractor.getData(entity.getId()).toList().toImmutable();
        return gameTranslator.translate(entity, genres, libraries);
    }

    @Override
    @SneakyThrows
    public Game getGameById(Id<Game> id) {
        final GameEntity entity = gameDao.queryForId(id.getId());
        if (entity == null) {
            throw new DataException("Couldn't find game with id: %d", id);
        }
        final ImmutableList<Genre> genres = getGenresByGame(entity);
        final ImmutableList<Library> libraries = getLibrariesByGame(entity);
        return gameTranslator.translate(entity, genres, libraries);
    }

    private ImmutableList<Genre> getGenresByGame(GameEntity game) throws SQLException {
        final ImmutableList<GenreEntity> entities = genreGameLinkDao.getGenresByGameId(game.getId());
        return entities.collect(genreTranslator::translate);
    }

    private ImmutableList<Library> getLibrariesByGame(GameEntity entity) throws SQLException {
        final ImmutableList<LibraryEntity> entities = libraryGameLinkDao.getLibrariesByGameId(entity.getId());
        return entities.collect(libraryTranslator::translate);
    }

    @Override
    @SneakyThrows
    public boolean hasGameForPath(Path path) {
        return gameDao.queryByPath(path) != null;
    }

    @Override
    @SneakyThrows
    public Library addLibrary(Path path, GamePlatform platform, String name) {
        final LibraryEntity entity = new LibraryEntity();
        entity.setPath(path.toString());
        entity.setPlatform(platform);
        entity.setName(name);

        LOG.debug("Inserting {}...", entity);
        if (libraryDao.create(entity) != 1) {
            throw new DataException("Error inserting library: %s", entity);
        }
        return libraryTranslator.translate(entity);
    }

    @Override
    @SneakyThrows
    public void deleteLibrary(Id<Library> id) {
        final int libraryId = id.getId();
        LOG.debug("Deleting library of id {}...", libraryId);
        if (libraryDao.deleteById(libraryId) != 1) {
            throw new DataException("Error deleting library by id: %s. Entry doesn't exist?", libraryId);
        }
        libraryGameLinkDao.deleteByLibraryId(libraryId);
    }

    @Override
    @SneakyThrows
    public ImmutableList<Library> getAllLibraries() {
        LOG.debug("Fetching all libraries...");
        final ImmutableList<LibraryEntity> entities = libraryDao.getAll();
        return entities.collect(libraryTranslator::translate);
    }

    @Override
    @SneakyThrows
    public Library getLibraryById(Id<Library> id) {
        final LibraryEntity entity = libraryDao.queryForId(id.getId());
        if (entity == null) {
            throw new DataException("Couldn't find library with id: %d", id);
        }
        return libraryTranslator.translate(entity);
    }

    @Override
    @SneakyThrows
    public boolean hasLibraryForPath(Path path) {
        return libraryDao.queryByPath(path) != null;
    }

    @Override
    @SneakyThrows
    public void addGameToLibraries(Game game, Iterable<Library> libraries) {
        for (Library library : libraries) {
            addGameToLibrary(game, library);
        }
    }

    private void addGameToLibrary(Game game, Library library) throws SQLException {
        LOG.debug("Adding game {} to library {}...", game, library);
        final GameEntity gameEntity = gameTranslator.translate(game);
        final LibraryEntity libraryEntity = libraryTranslator.translate(library);
        final LibraryGameLinkEntity linkEntity = new LibraryGameLinkEntity();
        linkEntity.setGame(gameEntity);
        linkEntity.setLibrary(libraryEntity);
        if (libraryGameLinkDao.create(linkEntity) != 1) {
            throw new DataException("Error adding game to library! Game=%s, Library=%s", game, library);
        }
    }

    @Override
    @SneakyThrows
    public ImmutableList<Genre> getAllGenres() {
        LOG.debug("Fetching all genres...");
        final ImmutableList<GenreEntity> entities = genreDao.getAll();
        return entities.collect(genreTranslator::translate);
    }

    @Override
    @SneakyThrows
    public ImmutableList<ExcludedPath> getAllExcludedPaths() {
        final ImmutableList<ExcludedPathEntity> entities = excludedPathDao.getAll();
        return entities.collect(excludedPathTranslator::translate);
    }

    @Override
    @SneakyThrows
    public boolean isPathExcluded(Path path) {
        return excludedPathDao.queryByPath(path) != null;
    }

    @Override
    @SneakyThrows
    public ExcludedPath addExcludedPath(Path path) {
        final ExcludedPathEntity entity = new ExcludedPathEntity();
        entity.setPath(path.toString());

        LOG.debug("Inserting {}...", entity);
        if (excludedPathDao.create(entity) != 1) {
            throw new DataException("Error inserting excluded path: %s", entity);
        }
        return excludedPathTranslator.translate(entity);
    }

    @Override
    @SneakyThrows
    public void deleteExcludedPath(Id<ExcludedPath> id) {
        final int excludedPathId = id.getId();
        LOG.debug("Deleting excluded path of id {}...", excludedPathId);
        if (excludedPathDao.deleteById(excludedPathId) != 1) {
            throw new DataException("Error deleting excluded path by id: %s. Entry doesn't exist?", excludedPathId);
        }
    }
}

File: gamedex-persistence/src/main/kotlin/com/github/ykrasik/gamedex/persistence/PersistenceServiceImpl.kt
package com.github.ykrasik.gamedex.persistence

import com.github.ykrasik.gamedex.common.debug
import com.github.ykrasik.gamedex.common.info
import com.github.ykrasik.gamedex.common.logger
import com.github.ykrasik.gamedex.datamodel.GamePlatform
import com.github.ykrasik.gamedex.datamodel.ImageData
import com.github.ykrasik.gamedex.datamodel.persistence.ExcludedPath
import com.github.ykrasik.gamedex.datamodel.persistence.Game
import com.github.ykrasik.gamedex.datamodel.persistence.Genre
import com.github.ykrasik.gamedex.datamodel.persistence.Library
import com.github.ykrasik.gamedex.datamodel.provider.GameData
import com.github.ykrasik.gamedex.persistence.config.PersistenceConfig
import com.github.ykrasik.gamedex.persistence.entity.*
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SchemaUtils.create
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.joda.time.DateTime
import java.math.BigDecimal
import java.nio.file.Path
import javax.inject.Inject
import javax.inject.Singleton

/**
 * @author Yevgeny Krasik
 */
@Singleton
class PersistenceServiceImpl @Inject constructor(private val config: PersistenceConfig) : PersistenceService {
    private val log by logger()

    init {
        log.debug { "Connection url: ${config.dbUrl}" }
        Database.connect(config.dbUrl, driver = "org.h2.Driver", user = "sa")

        // TODO: Use Db evolutions!!!
        log.debug("Creating db...")
        create(Libraries, Games, Genres, GameGenres, ExcludedPaths)
    }

    inner class GamesImpl : PersistenceService.Games {
        private val baseQuery = (Games innerJoin Libraries).slice(Games.withoutBlobs + Libraries.columns)

        override val all: List<Game> get() {
            val genres = gameGenres.oneToManyContext()
            log.info { "Fetching all games..." }
            return baseQuery.selectAll().map {
                val library = it.toLibrary()
                it.toGame(genres, library)
            }
        }

        override fun get(id: Int): Game {
            val genres = gameGenres.oneToManyContext(id)
            log.info { "Fetching game: $id..." }
            val row = baseQuery.select { Games.id.eq(id) }.first()
            val library = row.toLibrary()
            return row.toGame(genres, library)
        }

        override fun contains(path: Path): Boolean = !Games.selectBy { it.path.eq(path.toString()) }.empty()

        override fun add(gameData: GameData, path: Path, library: Library): Game {
            log.debug { "Inserting game: $gameData..." }
            val lastModified = DateTime.now()
            val id = Games.insert {
                it[Games.path] = path.toString()
                it[Games.name] = gameData.name
                it[Games.releaseDate] = gameData.releaseDate?.toDateTimeAtStartOfDay()
                it[Games.description] = gameData.description
                it[Games.criticScore] = gameData.criticScore?.let { BigDecimal(it) }
                it[Games.userScore] = gameData.userScore?.let { BigDecimal(it) }
                it[Games.metacriticUrl] = gameData.metacriticUrl
                it[Games.giantBombUrl] = gameData.giantBombUrl
                it[Games.thumbnail] = gameData.thumbnail?.rawData?.let { it.toBlob() }
                it[Games.poster] = gameData.poster?.rawData?.let { it.toBlob() }
                it[Games.lastModified] = lastModified
                it[Games.library] = library.id
            } get Games.id

            // Insert all new genres.
            // TODO: Room for optimization - do one inList bulk fetch for existing genres.
            val genres = gameData.genres.map { genres.getOrAdd(it) }

            // Link genres to game.
            gameGenres.add(id, genres)

            return Game(
                id = id,
                path = path,
                name = gameData.name,
                description = gameData.description,
                releaseDate = gameData.releaseDate,
                criticScore = gameData.criticScore,
                userScore = gameData.userScore,
                lastModified = lastModified,
                metacriticUrl = gameData.metacriticUrl,
                giantBombUrl = gameData.giantBombUrl,
                genres = genres,
                library = library
            )
        }

        override fun delete(id: Int) {
            log.info { "Deleting gameId=$id..." }
            require(Games.deleteWhere { Games.id.eq(id) } == 1) { "Failed to delete gameId=$id!" }

            val genres = gameGenres.oneToManyContext(id)[id]
            gameGenres.deleteByGame(id)

            // Delete any genres which were only linked to this game.
            val genresToDelete = genres.mapNotNull { genre ->
                if (gameGenres.isGenreLinked(genre.id)) {
                    null
                } else {
                    log.debug { "GenreId=${genre.id} was only linked to gameId=$id, deleting..." }
                    genre.id
                }
            }
            if (!genresToDelete.isEmpty()) {
                this@PersistenceServiceImpl.genres.deleteByIds(genresToDelete)
            }
        }

        override fun getThumbnail(id: Int): ImageData? {
            val result = Games.slice(Games.thumbnail).select { Games.id.eq(id) }.first()
            return result[Games.thumbnail]?.let { ImageData(it.bytes) }
        }

        override fun getPoster(id: Int): ImageData? {
            val result = Games.slice(Games.poster).select { Games.id.eq(id) }.first()
            return result[Games.poster]?.let { ImageData(it.bytes) }
        }

        fun getByLibrary(libraryId: Int): List<Int> {
            log.info { "Fetching games of library: $libraryId" }
            val ids = Games.select { Games.library.eq(libraryId) }.map {
                it[Games.id]
            }
            log.info { "Fetched ${ids.size} games." }
            return ids
        }

        fun deleteByLibrary(libraryId: Int) {
            log.info { "Deleting games by library: $libraryId" }
            val amount = Games.deleteWhere { Games.library.eq(libraryId) }
            log.info { "Deleted $amount games." }
        }
    }

    inner class GenresImpl : PersistenceService.Genres {
        override val all: List<Genre> get() = Genres.selectAll().map { it.toGenre() }

        fun add(name: String): Genre {
            log.info { "Inserting Genre: $name" }
            val id = Genres.insert {
                it[Genres.name] = name
            } get Genres.id
            val genre = Genre(id, name)
            log.info { "Inserted: $genre" }
            return genre
        }

        fun getByName(name: String): Genre? = Genres.selectBy { it.name.eq(name) }.firstOrNull()?.toGenre()

        fun getOrAdd(name: String): Genre = getByName(name) ?: add(name)

        fun deleteByIds(ids: List<Int>) {
            log.info { "Deleting genres: $ids" }
            val amount = Genres.deleteWhere { Genres.id.inList(ids) }
            log.info { "Deleted $amount genres." }
        }
    }

    private val gameGenres = object {
        fun add(gameId: Int, genres: List<Genre>) {
            GameGenres.batchInsert(genres, false) { genre ->
                log.debug { "Linking gameId=$gameId with genreId=${genre.id}..." }
                set(GameGenres.game, gameId)
                set(GameGenres.genre, genre.id)
            }
        }
//
//        fun add(gameId: Int, genre: Genre) {
//            log.debug { "Linking gameId=$gameId with genreId=${genre.id}..." }
//            GameGenres.insert {
//                it[GameGenres.game] = EntityID(gameId, Games)
//                it[GameGenres.genre] = EntityID(genre.id, Genres)
//            }
//        }

        fun deleteByGame(gameId: Int) {
            log.debug { "Un-linking all genres from gameId=$gameId..." }
            val amount = GameGenres.deleteWhere { GameGenres.game.eq(gameId) }
            log.info { "Un-linked $amount genres." }
        }

        fun isGenreLinked(genreId: Int): Boolean = !GameGenres.select { GameGenres.genre.eq(genreId) }.empty()

        fun oneToManyContext(id: Int = 0): OneToManyContext<Genre> {
            val join = GameGenres innerJoin Genres
            val query = if (id == 0) {
                log.info { "Fetching all gameGenres..." }
                join.selectAll()
            } else {
                log.info { "Fetching gameGenres for gameId=$id..." }
                join.select { GameGenres.game.eq(id) }
            }
            val mapping = query.groupBy(keySelector = { it[GameGenres.game] }, valueTransform = { it.toGenre() })
            return OneToManyContext(mapping)
        }
    }

    inner class LibrariesImpl : PersistenceService.Libraries {
        override val all: List<Library> get() = Libraries.selectAll().map { it.toLibrary() }

        override fun get(id: Int) = Libraries.select { Libraries.id.eq(id) }.first().toLibrary()

        override fun contains(path: Path) = !Libraries.select { Libraries.path.eq(path.toString()) }.empty()

        override fun add(path: Path, platform: GamePlatform, name: String): Library {
            log.info { "Inserting library: path=$path, platform=$platform, name=$name" }
            val id = Libraries.insert {
                it[Libraries.path] = path.toString()
                it[Libraries.platform] = platform
                it[Libraries.name] = name
            } get Libraries.id
            return Library(id, path, platform, name)
        }

        // Returns ids of games to be deleted.
        override fun delete(id: Int): List<Int> {
            val gameIds = games.getByLibrary(id)

            // TODO: onDelete Cascade does the exact same thing
            games.deleteByLibrary(id)

            log.info { "Deleting library: $id" }
            require(Libraries.deleteWhere { Libraries.id.eq(id) } == 1) { "Library doesn't exist: $id" }
            log.info { "Deleted." }
            return gameIds
        }
    }

    inner class ExcludedPathsImpl : PersistenceService.ExcludedPaths {
        override val all: List<ExcludedPath> get() {
            log.info { "Fetching all excludedPaths..." }
            return ExcludedPaths.selectAll().map { it.toExcludedPath() }
        }

        override fun contains(path: Path) = !ExcludedPaths.selectBy { it.path.eq(path.toString()) }.empty()

        override fun add(path: Path): ExcludedPath {
            log.info { "Inserting excludedPath: path=$path" }
            val id = ExcludedPaths.insert {
                it[ExcludedPaths.path] = path.toString()
            } get ExcludedPaths.id
            val excludedPath = ExcludedPath(id, path)
            log.info { "Inserted: $excludedPath" }
            return excludedPath
        }

        override fun delete(id: Int) {
            log.info { "Deleting excludedPath: id=$id..." }
            require(ExcludedPaths.deleteWhere { ExcludedPaths.id.eq(id) } == 1) { "ExcludedPath doesn't exist: $id" }
            log.info { "Deleted." }
        }
    }

    override val games = GamesImpl()
    override val genres = GenresImpl()
    override val libraries = LibrariesImpl()
    override val excludedPaths = ExcludedPathsImpl()
}
